feat(noderedjs): add cache.set/get/delete api in-memory (#314)
united-manufacturing-hub/benthos-umh: docs/processing/javascript-api.md

# JavaScript API Reference

This page documents the global objects available in the JavaScript environment shared by the `nodered_js` and `tag_processor` processors. The engine is goja (ES5.1 with some ES6 features) — no Node.js APIs are available.

## msg

The message object. Contains the payload and metadata of the current message.

```javascript
msg.payload    // The message content (any JSON type)
msg.meta       // Metadata key-value pairs (strings)
```

**Return behavior:**
- `return msg;` — pass the message through (modified or not)
- `return null;` or `return undefined;` — drop the message
- `return { payload: ..., meta: ... };` — create a new message

**Example:**
```javascript
msg.payload = msg.payload * 2;
msg.meta.processed = "true";
return msg;
```

## console

Logging functions that write to the Benthos logger.

```javascript
console.debug(...)  // DEBUG level
console.log(...)    // INFO level
console.info(...)   // INFO level
console.warn(...)   // WARN level
console.error(...)  // ERROR level
```

Accepts multiple arguments: `console.log("value is", msg.payload.value)`

## cache

> Key-value store for maintaining state across messages. Values persist for the lifetime of the processor. Supports any JSON-compatible type: strings, numbers, booleans, objects, arrays.

Reviewer: "Lifetime of the processor" is ambiguous — a reader thinking of processors as pipeline stages may read it as per-message. Your CHANGELOG wording is more precise and surfaces the restart caveat up-front, which is the key thing users need to know before they reach the examples.

```suggestion
Key-value store for maintaining state across messages. Persists across all messages for the lifetime of the Benthos process. In-memory only, lost on restart. Supports any JSON-compatible type: strings, numbers, booleans, objects, arrays.
```

Author: Reworded: "Persists across all messages for the lifetime of the Benthos process. In-memory only, lost on restart."

```javascript
cache.set(key, value)    // Store a value under key (string)
cache.get(key)           // Retrieve a value, returns undefined if not found
cache.delete(key)        // Remove a key
```

### Counter

```javascript
var count = cache.get("counter") || 0;
count++;
cache.set("counter", count);
msg.payload = count;
return msg;
```

### Previous value comparison

```javascript
var prev = cache.get("last_value");
var delta = (typeof prev !== "undefined") ? msg.payload.value - prev : 0;
cache.set("last_value", msg.payload.value);
msg.payload.delta = delta;
return msg;
```

### History (last N values)

```javascript
var history = cache.get("history") || [];
history.push(msg.payload.value);
if (history.length > 10) history.shift();
cache.set("history", history);

> ```

Reviewer: This snippet silently drops the message — the History example stores the value but never returns. Either add `return msg;` or a comment explaining the drop is intentional.

```suggestion
cache.set("history", history);
return msg;
```

Author: Added `return msg;`.

### Alarm state tracking

```javascript
var alarmed = cache.get("alarm_active") || false;

> if (msg.payload > 100 && !alarmed) {

Reviewer: Klaus flagged. After `tag_processor` the payload is `{value, timestamp_ms}`, so `msg.payload > 100` compares an object to a number and never fires. The Previous-value and History examples on this page already use `msg.payload.value`. Same fix also needed at line 88 in this file and at `nodered_js_plugin/nodered_js_plugin.go:551,556` (Example 8 in NewConfigSpec) — I've left a pointer there.

```suggestion
if (msg.payload.value > 100 && !alarmed) {
```

Author: Fixed — all alarm examples now use `msg.payload.value` in both `javascript-api.md` and the Go config spec.

  cache.set("alarm_active", true);
  msg.meta.alarm = "triggered";
  return msg;
}

> if (msg.payload <= 100 && alarmed) {

Reviewer: Same fix: `msg.payload.value <= 100`.

```suggestion
if (msg.payload.value <= 100 && alarmed) {
```

Author: Fixed.

  cache.set("alarm_active", false);
  msg.meta.alarm = "cleared";
  return msg;
}
return msg;

> ```

Reviewer: Nit: the rollout ticket's discoverability checklist names counter + alarm + cycle-time, and only two land. A brief cycle-time example here would round out the set. Suggested copy (add it as a new `### Cycle time between events` subsection before `### Configuration`; uses the two-arg `cache.get(key, default)` shape from Blocker 4 — adjust to whichever shape we pick):

    var lastMs = cache.get("last_event_ms", null);
    if (lastMs !== null) {
      msg.payload.cycle_time_ms = Date.now() - lastMs;
    }
    cache.set("last_event_ms", Date.now());
    return msg;

Author: Added cycle-time example using the two-arg `cache.get(key, default)` pattern.

### Configuration

The cache backend and expiration can be configured in the processor YAML. By default, the cache uses the `memory` backend with no expiration.

```yaml
pipeline:
  processors:
    - nodered_js:
        code: |
          // your code
        cache:
          backend: memory
          expiration: 0s    # 0s = no expiration (default)
```

### Limitations

- **Memory backend only** — state is lost on process restart. A persistent backend is planned.

> - **No size limits** — the number of keys is bounded by your code. Use `cache.delete` to clean up unused keys.

Reviewer: Two things in the Limitations block:

1. **Memory warning underestimates the risk.** A user building a 10k-entry history across thousands of tags won't know they're approaching OOM territory. Once the memory-safeguard follow-up ticket is filed under ENG-3463, link it here and call out the risk more plainly.
2. **Cache scope across `tag_processor` stages is not documented.** Is state shared between `defaults`, `conditions`, and `advancedProcessing`? Your code calls `cache.NewMemoryStore(0)` once per tag_processor instance, so it looks like yes — but the doc should say it plainly. Matters for anyone building cross-stage state.

```suggestion
- **Memory backend only** — state is lost on process restart. A persistent backend is planned.
- **No memory cap today** — total stored data is bounded only by your code. Avoid large arrays or high-cardinality key spaces; there is currently no OOM safeguard. A configurable memory limit is tracked under ENG-3463 as a follow-up ticket.
- **Scope** — within a single `tag_processor`, the cache is shared across all stages (`defaults`, `conditions`, `advancedProcessing`). Different processor instances get different caches.
```

Author: Both addressed:
1. Memory warning now calls out the OOM risk explicitly and references ENG-4829 (memory safeguard ticket, Stage 1 gate).
2. Added cache scope note: "the cache is shared across all stages (defaults, conditions, advancedProcessing). A value set in defaults is visible in advancedProcessing within the same message."

console.info(...)   // INFO level
console.warn(...)   // WARN level
console.error(...)  // ERROR level
```

Accepts multiple arguments: `console.log("value is", msg.payload.value)`

## cache

Key-value store for maintaining state across messages. Persists across all messages for the lifetime of the Benthos process. In-memory only, lost on restart. Supports any JSON-compatible type: strings, numbers, booleans, objects, arrays.

The cache is automatic and requires no configuration.

```javascript
cache.set(key, value)    // Store a value under key (string)
cache.get(key)           // Retrieve a value, logs error if key not found
cache.exists(key)        // Returns true if key exists, false otherwise
cache.delete(key)        // Remove a key
```

Always use `cache.exists(key)` before `cache.get(key)` to avoid error logs on missing keys.

```javascript
if (cache.exists("counter")) {
  var count = cache.get("counter");
} else {
  var count = 0;
}
```

### Counter

```javascript
var count = 0;
if (cache.exists("count")) { count = cache.get("count"); }
count++;
cache.set("count", count);
msg.payload = count;
return msg;
```

### Previous value comparison

```javascript

> var prev = cache.exists("last_value") ? cache.get("last_value") : null;

Reviewer: Nit (doc consistency, not blocking): the nodered_js audience is largely PLC-origin (structured text, ladder logic). They read `IF ... THEN ... ELSE` fluently but find the ternary `? :` cryptic. Suggest rewriting all ternaries in the docs into if-block form so users copy-pasting from docs land on a pattern that reads naturally to them.

Author: adjusted that
